server/app/app.py

```diff
@@ -30,13 +193,25 @@ def create_app(self):
         # Config
         SECRET_KEY = os.environ.get("CXG_SECRET_KEY", default="SparkleAndShine")
         self.app.config.update(SECRET_KEY=SECRET_KEY)
-        self.app.config.update(SCRIPTS=[])
 
-        resources = get_api_resources()
-        self.app.register_blueprint(webapp.bp)
-        self.app.register_blueprint(resources.blueprint)
-        self.app.add_url_rule("/", endpoint="index")
+        self.app.register_blueprint(webbp)
+
+        api_version = "/api/v0.2"
+        if data:
+            bp_api = Blueprint("api", __name__, url_prefix=api_version)
+            resources = get_api_resources(bp_api)
+            self.app.register_blueprint(resources.blueprint)
+            self.app.add_url_rule("/", "index", index)
+
+        if app_config.dataroot:
+            bp_api = Blueprint("api_dataset", __name__, url_prefix="/<dataset>" + api_version)
+            resources = get_api_resources(bp_api)
+            self.app.register_blueprint(resources.blueprint)
+            self.app.add_url_rule("/<dataset>/", 'dataset_index', dataset_index)
+            self.app.add_url_rule("/<dataset>/static/<path:therest>", "static_redirect", static_redirect)
+            self.app.add_url_rule("/<dataset>/favicon.png", "favicon_redirect", favicon_redirect)
 
-    def attach_data(self, data, title="Demo", about=""):
-        self.app.config.update(DATASET_TITLE=title, ABOUT_DATASET=about)
         self.app.data = data
```

Comment:
name of the instance variable is quite ambiguous.    Maybe `default_data_adaptor`?

Reply:
Removing data, and refactoring like we talked about.